File: is_symmetric.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isSymmetric(root):
    if not root:
        return True
    return isSame(root.left, root.right)

def isSame(leftroot, rightroot):
    if (not leftroot and rightroot) or (leftroot and not rightroot) or (leftroot and rightroot and leftroot.val != rightroot.val):
        return False
    if leftroot and rightroot:
        return isSame(leftroot.left, rightroot.right) and isSame(leftroot.right, rightroot.left)
    return True

File: IsSymmetric.java
class TreeNode {
    
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class IsSymmetric {
    public boolean isSymmetric(TreeNode root) {
        if (root == null)
            return true;
        return isSame(root.left, root.right);
    }

    private boolean isSame(TreeNode leftroot, TreeNode rightroot) {
        if ((leftroot == null && rightroot != null) || (leftroot != null && rightroot == null)
                || (leftroot != null && rightroot != null && leftroot.val != rightroot.val))
            return false;
        if (leftroot != null && rightroot != null)
            return isSame(leftroot.left, rightroot.right) && isSame(leftroot.right, rightroot.left);
        return true;
    }
}
